File: src/server/lib/markdown-parser.test.ts
import { describe, expect, it } from "vitest";
import { parseMarkdownFrontmatter, truncateBody } from "./markdown-parser.js";

describe("parseMarkdownFrontmatter", () => {
  it("parses frontmatter and body", () => {
    const content = `---
title: Hello
tags:
  - a
  - b
---
# Body content

Some text here.`;

    const result = parseMarkdownFrontmatter(content);
    expect(result.frontmatter).toEqual({ title: "Hello", tags: ["a", "b"] });
    expect(result.body).toBe("# Body content\n\nSome text here.");
  });

  it("returns empty frontmatter when none present", () => {
    const result = parseMarkdownFrontmatter("Just a plain body.");
    expect(result.frontmatter).toEqual({});
    expect(result.body).toBe("Just a plain body.");
  });

  it("trims whitespace from body", () => {
    const content = `---
key: value
---

  Some body
`;
    const result = parseMarkdownFrontmatter(content);
    expect(result.body).toBe("Some body");
  });

  it("handles empty content", () => {
    const result = parseMarkdownFrontmatter("");
    expect(result.frontmatter).toEqual({});
    expect(result.body).toBe("");
  });
});

describe("truncateBody", () => {
  it("returns body unchanged when under max lines", () => {
    const body = "line1\nline2\nline3";
    expect(truncateBody(body, 10)).toBe(body);
  });

  it("returns body unchanged when exactly at max lines", () => {
    const body = "1\n2\n3";
    expect(truncateBody(body, 3)).toBe(body);
  });

  it("truncates body over max lines and appends ellipsis", () => {
    const body = "1\n2\n3\n4\n5";
    expect(truncateBody(body, 3)).toBe("1\n2\n3\n...");
  });

  it("defaults to 10 lines", () => {
    const lines = Array.from({ length: 15 }, (_, i) => `line${i + 1}`);
    const result = truncateBody(lines.join("\n"));
    expect(result.split("\n")).toHaveLength(11); // 10 lines + "..."
    expect(result.endsWith("...")).toBe(true);
  });
});

File: server/lib/markdownparser_test.go
package lib_test

import (
	"testing"

	"fieldstation/lib"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestParseMarkdown_NoFrontmatter(t *testing.T) {
	content := "# Hello World\n\nThis is a markdown document."
	doc, err := lib.ParseMarkdownFrontmatter(content)
	require.NoError(t, err)
	assert.Empty(t, doc.Frontmatter, "frontmatter should be empty when no --- fence")
	assert.Equal(t, content, doc.Body)
}

func TestParseMarkdown_WithFrontmatter(t *testing.T) {
	content := "---\ntitle: My Document\nauthor: Alice\n---\n# Body Content\n\nSome text here."
	doc, err := lib.ParseMarkdownFrontmatter(content)
	require.NoError(t, err)
	assert.Equal(t, "My Document", doc.Frontmatter["title"])
	assert.Equal(t, "Alice", doc.Frontmatter["author"])
	assert.Equal(t, "# Body Content\n\nSome text here.", doc.Body)
}

func TestParseMarkdown_FrontmatterTypes(t *testing.T) {
	content := "---\ncount: 42\nenabled: true\ntags:\n  - go\n  - yaml\n---\nBody text."
	doc, err := lib.ParseMarkdownFrontmatter(content)
	require.NoError(t, err)
	assert.Equal(t, 42, doc.Frontmatter["count"])
	assert.Equal(t, true, doc.Frontmatter["enabled"])
	tags, ok := doc.Frontmatter["tags"].([]any)
	require.True(t, ok, "tags should be a slice")
	assert.Equal(t, []any{"go", "yaml"}, tags)
	assert.Equal(t, "Body text.", doc.Body)
}

func TestParseMarkdown_EmptyFrontmatter(t *testing.T) {
	content := "---\n---\n"
	doc, err := lib.ParseMarkdownFrontmatter(content)
	require.NoError(t, err)
	assert.Empty(t, doc.Frontmatter, "frontmatter should be empty for empty fence")
	assert.Equal(t, "", doc.Body)
}

func TestSerializeMarkdown_WithFrontmatter(t *testing.T) {
	original := "---\ntitle: Round Trip\n---\nBody content here."
	doc, err := lib.ParseMarkdownFrontmatter(original)
	require.NoError(t, err)

	serialized := lib.SerializeMarkdown(doc)

	// Re-parse to verify round-trip
	doc2, err := lib.ParseMarkdownFrontmatter(serialized)
	require.NoError(t, err)
	assert.Equal(t, doc.Frontmatter["title"], doc2.Frontmatter["title"])
	assert.Equal(t, doc.Body, doc2.Body)
}

func TestSerializeMarkdown_EmptyFrontmatter(t *testing.T) {
	doc := lib.FrontmatterDoc{
		Frontmatter: map[string]any{},
		Body:        "Just a body, no frontmatter.",
	}
	result := lib.SerializeMarkdown(doc)
	assert.Equal(t, "Just a body, no frontmatter.", result, "should return body unchanged when frontmatter is empty")
	assert.NotContains(t, result, "---", "should not contain --- fence when frontmatter is empty")
}

func TestSerializeMarkdown_NilFrontmatter(t *testing.T) {
	doc := lib.FrontmatterDoc{
		Frontmatter: nil,
		Body:        "No frontmatter at all.",
	}
	result := lib.SerializeMarkdown(doc)
	assert.Equal(t, "No frontmatter at all.", result)
}

func TestParseMarkdown_BodyTrimmedLeadingNewline(t *testing.T) {
	// After closing ---, there may be a newline before the body
	content := "---\nkey: val\n---\n\nParagraph after blank line."
	doc, err := lib.ParseMarkdownFrontmatter(content)
	require.NoError(t, err)
	// Body should be trimmed of the leading newline
	assert.Equal(t, "Paragraph after blank line.", doc.Body)
}

func TestTruncateBody_ShortBody(t *testing.T) {
	body := "line1\nline2\nline3"
	result := lib.TruncateBody(body, 10)
	assert.Equal(t, body, result, "short body should be returned unchanged")
}

func TestTruncateBody_LongBody(t *testing.T) {
	lines := make([]string, 15)
	for i := range lines {
		lines[i] = "line"
	}
	body := "line\nline\nline\nline\nline\nline\nline\nline\nline\nline\nline\nline\nline\nline\nline"
	result := lib.TruncateBody(body, 10)
	assert.Equal(t, "line\nline\nline\nline\nline\nline\nline\nline\nline\nline\n...", result)
}

func TestTruncateBody_DefaultMaxLines(t *testing.T) {
	// Default is 10 lines
	body := "1\n2\n3\n4\n5\n6\n7\n8\n9\n10\n11"
	result := lib.TruncateBody(body, 10)
	assert.Equal(t, "1\n2\n3\n4\n5\n6\n7\n8\n9\n10\n...", result)
}
